pkg/icons/icons.go

```diff
@@ -0,0 +1,144 @@
+package icons
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"helm.sh/helm/v3/pkg/repo"
+)
+
+const (
+	partnerFilePath     = "index.yaml"
+	partnerDownloadPath = "assets/logos"
+)
+
+// PackageIconOverride is a struct to hold the package name, path, and icon to override
+type PackageIconOverride struct {
+	Name string
+	Path string
+	Icon string
+}
+
+// PackageIconList is a list of PackageIconOverride
+type PackageIconList []PackageIconOverride
+
+// PackageIconMap is a map of PackageIconOverride
+type PackageIconMap map[string]PackageIconOverride
+
+// ParsePackageToOverride will convert a Package from a PackageList to a PackageIcon for a PackageIconList
+func ParsePackageToOverride(name, path, icon string) PackageIconOverride {
+	return PackageIconOverride{
+		Name: name,
+		Path: path,
+		Icon: icon,
+	}
+}
+
+// CheckFilesStructure will check if index.yaml and assets/logos exist
+// will log and exit if the file structure does not exist
+func CheckFilesStructure() {
+	if _, err := os.Stat(partnerFilePath); os.IsNotExist(err) {
+		fmt.Printf("File not found: %s\n", partnerFilePath)
+		logrus.Fatalf("File not found: %s\n", partnerFilePath)
+	}
+	_, err := os.Stat(partnerDownloadPath)
+	if os.IsNotExist(err) {
+		logrus.Fatalf("Directory not found: %s\n", partnerDownloadPath)
+	}
+}
+
+// CheckForOverrideConditions will check if the package has an upstream.yaml file and if the iconURL is a local file
+func CheckForOverrideConditions(packagePath string, iconURL string) bool {
+	var err error
+	if os.Stat(fmt.Sprintf("%s/upstream.yaml", packagePath)); os.IsNotExist(err) {
+		return false
+	}
+	if iconURL != "" {
+		if strings.HasPrefix(iconURL, "file://assets/logos") {
+			return true
+		}
+	}
+	return false
+}
+
+// ConvertYamlForIconOverride will change the metade icon URL to a local icon path for the index.yaml
+func ConvertYamlForIconOverride(helmIndexYaml *repo.IndexFile, packageIconList PackageIconList) {
+	for _, pkg := range packageIconList {
+		for _, pk := range helmIndexYaml.Entries[pkg.Name] {
+			pk.Metadata.Icon = pkg.Icon
+		}
+	}
+}
+
+// TestIconsAndIndexYaml will if Icons and IndexYaml have matching paths
+func TestIconsAndIndexYaml(packageIconList PackageIconList, helmIndexFile *repo.IndexFile) error {
+	var err error
+	for _, pkg := range packageIconList {
+		// Check if the icon has the correct path prefix
+		if !strings.HasPrefix(pkg.Icon, "file://") {
+			continue // Skip if the icon is not a local file
+		}
+		iconPath := strings.TrimPrefix(pkg.Icon, "file://")
+		// Check if icon exists in the path
+		_, openIconErr := os.Stat(iconPath)
+		if os.IsNotExist(openIconErr) {
```

Comment:
We could collapse this line and the above line into one. Also, we could use an `Exists` function here.

Reply:
This I disagree. 
I find it harder to read the combination of `variable declaration` +  `if statements` in go. 
However, when I was reading this block of code, I realized I was treating the errors the wrong way. 
Please take a look at that.